# HW10 
## Zachary Miller

In [1]:
import numpy as np
import random
random.seed(42)

### 1a
We are given the below parameters and asked to draw the neural network they represent along with the equations for each layer. This is done by hand below.

In [2]:
w2=np.array([[ 0.336067, -0.322224],[-0.700076, 0.00861628]])
b2=np.array([ 4.17819,-6.55492]).reshape(2,1)
w3= np.array([[ -106.923, -1936.34],[-36.9631, 1949.76]])    
b3=np.array([113.353, 27.4203]).reshape(2,1)
w4=np.array([1138.29, 2871.45])
b4=np.array([-1138.24])

### 1b
We are now asked to calculate the outputs for five different possible inputs, as well as for the input $[1.5;1,7]$. We are also asked to show the intermediate neuron values for the last input. Doing this below...

In [3]:
def sigmoid(x):
    sig=1.0/(1.0+np.exp(-x));
    return sig

# Calculate the output of each layer as a funciton of the first layer
# before and after the activation function
def calc_output(x, verbose=False):
    z2 = w2@x+b2
    a2 = sigmoid(z2)
    z3 = w3@a2+b3
    a3 = sigmoid(z3)
    z4 = w4@a3+b4
    a4 = 1*z4
    
    if verbose==True:
        print(f'z2:\n{z2}')
        print(f'\na2:\n{a2}')
        print(f'\nz3:\n{z3}')
        print(f'\na3:\n{a3}')
        print(f'\nz4:\n{z4}')
        print(f'\na4:\n{a4}')
    return a4

test_vals = np.random.rand(2,5)+1
print(f'Test Values:\n{test_vals}\n')

outputs = []
for i in range(5):
    test_val = test_vals[:,i].reshape(2,1)
    output = calc_output(test_val)
    outputs.append(output)
    print(f'Output for {test_vals[:,i]}:\n{output}\n')
    
print(f'[1.5;1.7] input with steps:\n')
output = calc_output(np.array([[1.5],[1.7]]),verbose=True)
outputs.append(output)
test_vals = np.hstack((test_vals, np.array([[1.5],[1.7]])))

Test Values:
[[1.07657994 1.82405421 1.24361556 1.02127347 1.56959945]
 [1.31792961 1.30440459 1.0474012  1.36628162 1.17881533]]

Output for [1.07657994 1.31792961]:
[0.22341123]

Output for [1.82405421 1.30440459]:
[-0.28575606]

Output for [1.24361556 1.0474012 ]:
[-0.15908463]

Output for [1.02127347 1.36628162]:
[0.33667149]

Output for [1.56959945 1.17881533]:
[-0.24985876]

[1.5;1.7] input with steps:

z2:
[[ 4.1345097 ]
 [-7.59038632]]

a2:
[[9.84241784e-01]
 [5.05030627e-04]]

z3:
[[ 7.13700473]
 [-7.97563897]]

a3:
[[9.99205502e-01]
 [3.43617085e-04]]

z4:
[0.13230966]

a4:
[0.13230966]


### 1c
Now we are asked to confirm that this network is the same as the function $y=(x_2-x_1)/x_1$. Doing this below...

In [4]:
# Define the function that we are given
def func(x):
    y = (x[1]-x[0])/x[0]
    return y

print('Expected output for given function:')
for i in range(6):
    test_val = test_vals[:,i].reshape(2,1)
    func_output = func(test_val)
    net_output = calc_output(test_val)
    print(f'Input:\n{test_val}')
    print(f'Function output:\n{func_output}')
    print(f'Network output:\n{net_output}\n')

Expected output for given function:
Input:
[[1.07657994]
 [1.31792961]]
Function output:
[0.22418184]
Network output:
[0.22341123]

Input:
[[1.82405421]
 [1.30440459]]
Function output:
[-0.28488716]
Network output:
[-0.28575606]

Input:
[[1.24361556]
 [1.0474012 ]]
Function output:
[-0.15777734]
Network output:
[-0.15908463]

Input:
[[1.02127347]
 [1.36628162]]
Function output:
[0.33782151]
Network output:
[0.33667149]

Input:
[[1.56959945]
 [1.17881533]]
Function output:
[-0.2489706]
Network output:
[-0.24985876]

Input:
[[1.5]
 [1.7]]
Function output:
[0.13333333]
Network output:
[0.13230966]



Looking at the above ouput, we can see that the network does indeed approximate the given funciton. 

### 2a
We are asked to write the explicit equations for the network given by the values below and draw its architecture, which is done by hand below. We are then asked to calculate the output and cost of the given training point and label, showing all intermediate steps. Doing this below...

In [5]:
w2=np.array([[0.2,0.2],[-0.4,0.2],[0.1,-0.1]])
b2=np.array([0.2,0.4,0.4]).reshape(3,1)
w3=np.array([-0.1, 0.2, 0.2])
b3=np.array([0.2])

In [14]:
def calc_output(x, verbose=False):
    z2 = w2@x+b2
    a2 = sigmoid(z2)
    z3 = w3@a2+b3
    a3 = sigmoid(z3)
    
    if verbose==True:
        print(f'z2:\n{z2}')
        print(f'\na2:\n{a2}')
        print(f'\nz3:\n{z3}')
        print(f'\na3:\n{a3}')
        return [z2,a2,z3,a3]
        
    return a3

def calc_loss(output, label):
    loss=0.5*np.square(label-output)
    return loss


print('Output for [3;2]:')
x = np.array([[3],[2]])
y = 1
output = calc_output(x, verbose=True)
loss=calc_loss(output[3], y)
print(f'\nCost: {loss}')

Output for [3;2]:
z2:
[[ 1.2]
 [-0.4]
 [ 0.5]]

a2:
[[0.76852478]
 [0.40131234]
 [0.62245933]]

z3:
[0.32790186]

a3:
[0.58124878]

Cost: [0.08767629]


### 2b
We are now asked to write out the backpropogation algorithm and use it to calculate the gradient of the cost function with respect to each set of weights $\vec{w}^l$ and biases $\vec{b}^l$. To do this, we start by finding the gradient of the output layer with respect the input values of the layer, which is given by $\frac{\partial C}{\partial \vec{z}^L}$. We will denote this quantity as $\delta^L$, and use the chain rule to express this as 
$$\delta^L=\frac{\partial C}{\partial \vec{z}^L}=\frac{\partial C}{\partial \vec{a}^L} \frac{\partial \vec{a}^L}{\partial \vec{z}^L}$$

Looking at the above equation, we see that $\frac{\partial C}{\partial \vec{a}^L}$ is easily calculated as $\frac{d}{d\vec{a}^L}(\frac{1}{2}(y-a^L)^2)=y-a^L$, which we denote as $\nabla_{\vec{a}^L}C$. We can also see that $\frac{\partial \vec{a}^L}{\partial \vec{z}^L}$ is simply the derivative of the activation function with respect to $z^L$, which is given by $\frac{d}{dz^L}=(1+e^{-z^{L}})^{-1}=e^{-z^L}(1+e^{-z^L})^{-2}$, which we will denote as $\sigma '(\vec{z}^L)$. This gives us a very clean vector form of the above expression,
$$\vec{\delta}^L=\nabla_{\vec{a}^L}C\odot\sigma '(\vec{z}^L)$$


We started with the last layer since we can calculate the derivative of the cost function with respect to the input of each neuron of the last layer directly. But what about earlier layers? For these, we need think about the effect of all layers after the layer we are adjusting as well, which makes the problem quite a bit harder. However, using the chain rule, we can show that the gradient of each layer can be expressed in terms of the the inputs of the current layer, the weights going to the next layer, and the $\delta$ of the next layer, given by $\vec{\delta}^l=(W^{l+1})^T\vec{\delta^{l+1}}\odot\sigma '(\vec{z}^l)$. Using this expression along with the one we derived for the output layer, we can build our way backward though each layer one at a time, hence the name backpropogation!

However, we still need to know how to adjust our weights and biases for a given layer using $\vec{\delta}^l$. We can again use the chain rule, as is done in the course notes equation 7.13 and 7.14, to get 
$$\nabla_{\vec{b}^l}C=\vec{\delta}^l$$
$$\nabla_{W^l}C=\vec{\delta}^l(\vec{a}^{l-1})^T$$

Now we have explicit equations we need to calculate the gradient of the network with resect to all the weights and biases. Doing this below...

In [ ]:
def sigma_prime(x):
    np.exp(-x)/np.square(1+np.exp(-x))

def calc_grad(x, y, net_vals):
    z2, a2, z3, a3 = net_vals[0], net_vals[1], net_vals[2], net_vals[3]
    delta_3 = (y-a3)*sigma_prime(z3)
    delta_2 = (w3.T@delta_3)*simga_prime(z2)
    
    # now calculate the gradients for weights and biases of layers 2 and 3
    

In [15]:
a = np.array([1,2,3]).reshape(3,1)

In [17]:
2-a

array([[ 1],
       [ 0],
       [-1]])